photon-test/src/main/scala/com/linkedin/photon/ml/test/TestTemplateWithTmpDir.scala

```diff
@@ -35,7 +35,7 @@ trait TestTemplateWithTmpDir {
 }
 
 private object TestTemplateWithTmpDir {
-  private val _tmpDirThreadLocal: ThreadLocal[String] = new ThreadLocal[String] {
+  private def _tmpDirThreadLocal: ThreadLocal[String] = new ThreadLocal[String] {
```

Comment:
In some cases we may call getTmpDir a few times for the same test, for example, in a test with ```DataProvider``` that provides multiple parameters and thus multiple run of the same test. In such cases, different runs of the same test would have to share the same tmpDir (because it's defined as a val of a singleton) and exceptions like "directory *** already exists" may arise.

In such case, call by name might be a better choice than call by value.

For ```ThreadLocal```, it might still be useful in case two threads trying to use the same directory, although the chance for that is very slim given the way we name the tmp directory.

But I agree that we can remove ThreadLocal from this function.